releasenotes/notes/cached-method-87b5d878f585ca92.yaml

```diff
@@ -0,0 +1,13 @@
+---
+features:
+  - |
+    Adds caching of transpiled circuit generation to :class:`.BaseExperiment`
```

Comment:
If caching the transpile is the main purpose, I feel general cache decorator is not necessary. Especially coupling to options is my concern, because it just indirectly evaluates equality of experiment circuits. Instead, we can just cache transpiled circuit in the dictionary keyed on `hash(tuple(circuits))`.

Reply:
Another option would be to not put any caching in base experiment and leave it up to sub experiments, but they will have to override the `_transpiled_circuits` method to add the caching.

Reply:
Yeah the problem is RB (and experiment with similar random circuit) with unfixed seed. Delegating cache to sub experiment may work, but we need to write transpile method just for it, such as

```python
@cache_method
def _transpiled_circuits(self):
    return super()._transpiled_circuits()
```

which looks bit awkward.

Reply:
@chriseclectic @nkanazawa1989 
Even when running RB, we may often prefer to cache. Sometimes yes and sometimes not, depending on the user.

I suggest to add an experiment option to all the experiments, named `cache_methods`. Default value is `True`. `cached_method` will cache only if the flag is set to `True`.

With this solution, the decorator will remain at the base class level.